src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -1457,13 +1459,25 @@ private IEnumerable<TypeInferenceInfo> InferTypeInRecursivePattern(RecursivePatt
                 // we have  { X: ... }.  The type of ... is whatever the type of 'X' is in its
                 // parent type.  So look up the parent type first, then find the X member in it
                 // and use that type.
-                if (child == subpattern.Pattern &&
-                    // PROTOTYPE(extended-property-patterns) ExpressionColon
-                    subpattern.NameColon != null)
+                if (child == subpattern.Pattern)
+                {
+                    if (subpattern.NameColon != null)
+                    {
+                        return InferTypeInSubpatternCore(subpattern.NameColon.Name);
+                    }
+                    else if (subpattern.ExpressionColon is { Expression: MemberAccessExpressionSyntax memberAccess })
+                    {
+                        return InferTypeInSubpatternCore(memberAccess.Name);
+                    }
+                }
+
+                return SpecializedCollections.EmptyEnumerable<TypeInferenceInfo>();
+
+                IEnumerable<TypeInferenceInfo> InferTypeInSubpatternCore(ExpressionSyntax expression)
                 {
                     var result = ArrayBuilder<TypeInferenceInfo>.GetInstance();
```

Comment:
```suggestion
                    using var result = TemporaryAray<TypeInferenceInfo>.Empty;
```

Reply:
TemporaryArray is preferred when the expected case is a small number of results (4 or less).